Automattic/wp-calypso
# [GB][E2E] Add `getTestAccountByFeature` account-selection abstraction for E2E tests

Follow up to: https://github.com/Automattic/wp-calypso/pull/61349.
Project: p9oQ9f-13V-p2.

### Changes proposed in this Pull Request

This PR aims to abstract the WP account selection logic for our E2E tests in order to make it easier to support more complex scenarios where a specific site is selected based on external data (passed by env var / TC). The original goal was to make it trivial to support testing on AT - with the flip of a switch. I've done that by adding a new `TEST_ON_ATOMIC` TC param and env var that is then passed to a new function that processes this criteria (and others, more about it below) and returns the correct site (in this case, the AT site if the var is set -- and the AT site to be returned might even change based on whether or not the feature has a flag (i.e is Gutenberg `edge` or `stable`?).


#### Introducing the `getTestAccountByFeature` function

This PR adds a helper function to abstract the selection of an accountName for a test. It allows for a declarative way to define and select WP account names for each feature or combination of features (in this specific case, the only "features" available are `gutenberg` and `coblocks`, but we could extend this concept later to selectively select sites for more components or combo of components).

To illustrate, the function simplifies this:

```js
let accountName: string;
if ( envVariables.COBLOCKS_EDGE ) {
	accountName = 'coBlocksSimpleSiteEdgeUser';
} else if ( envVariables.GUTENBERG_EDGE ) {
	accountName = 'gutenbergSimpleSiteEdgeUser';
} else {
	accountName = 'gutenbergSimpleSiteUser';
}
```

Into: 

```js
const accountName = getTestAccountByFeature( {
	coblocks: envVariables.COBLOCKS_EDGE ? 'edge' : undefined,
	gutenberg: envVariables.GUTENBERG_EDGE ? 'edge' : 'stable',
	siteType: envVariables.TEST_ON_ATOMIC ? 'atomic' : 'simple',
} );
```

@worldomonation is working on refactoring some editor classes to support AT. These classes will feed from the account name returned by the new function and the new `TEST_ON_ATOMIC` env var to decide what implementation to use (we should assume AT doesn't use Gutenframe, see: . This is the main difference between the editor on simple vs Atomic at the moment). The changes by @worldomonation **shouldn't prevent this PR from being merged**. The changes here shouldn't break anything, they just allow us to "flip the switch" later by creating new builds to run these tests on AT.

Once these changes are in, we can then create a new build that will pass the `TEST_ON_ATOMIC` var and all the tests that we refactored to use the `getTestAccountByFeature` for the account-selection logic will start selecting the right AT site to test with (provided the actual criteria is present for it!).

From then on, it'll be a matter of seeing what errors and fixing to make them work in AT. Some tests might have to be skipped (we can use the `skipItIf` helper to skip if `TEST_ON_ATOMIC` is set) because they might test something that's too tied to the Gutenframe/Calypso, and that's fine.

### Next Steps

- Add AT-specific Gutenberg upgrade builds in TC that will set the `TEST_ON_ATOMIC` env variable automatically.

### Later

We could explore expanding the concept introduced by this function to conditionally return other artifacts based on different testing criteria. For example, it could return a more complex object with boolean flags in addition to the account name or anything else that needs to be conditional to external vars. Maybe it could be used to also return different components or test implementations based on the env. TBD.


### TODO

- [x] Convert tests that require a test account and have any kind of selection logic around it to use the `getTestAccountByFeature` function instead and define the conditions (criteria) declaratively as needed (in progress)
- [x] Add tests (in progress)
- [x] Add a `AT_TEST` (naming is hard!) parameter to TC and to the `envVariables` in `calypso-e2e`. This will be passed to `getTestSite` to get a proper account based on its presence (besides `GUTENBERG_EDGE`).
- [x] Add jsdoc for `getTestAccountByFeature`
- [x] Fix lint errors
- [x] Rebase and fix conflicts


#### Testing instructions

* Tests should pass



Related to https://github.com/Automattic/wp-calypso/pull/61724, https://github.com/Automattic/wp-calypso/pull/61932
